hls-test-utils/src/Test/Hls.hs

```diff
@@ -9,38 +9,116 @@ module Test.Hls
+import           Test.Tasty                        hiding (Timeout)
 import           Test.Tasty.ExpectedFailure
 import           Test.Tasty.Golden
 import           Test.Tasty.HUnit
 import           Test.Tasty.Ingredients.Rerun
-import           Test.Tasty.Runners
-import           Test.Tasty.Runners.AntXML
 
--- | ingredient: xml runner writes json file of test results (https://github.com/ocharles/tasty-ant-xml/blob/master/Test/Tasty/Runners/AntXML.hs)
---               rerunningTests allow rerun of failed tests (https://github.com/ocharles/tasty-rerun/blob/master/src/Test/Tasty/Ingredients/Rerun.hs)
+-- | Run 'defaultMainWithRerun', limiting each single test case running at most 10 minutes
 defaultTestRunner :: TestTree -> IO ()
-defaultTestRunner =
-  defaultMainWithIngredients
-    [antXMLRunner, rerunningTests [listingTests, consoleTestReporter]]
+defaultTestRunner = defaultMainWithRerun . adjustOption (const $ mkTimeout 600000000)
 
 gitDiff :: FilePath -> FilePath -> [String]
 gitDiff fRef fNew = ["git", "diff", "--no-index", "--text", "--exit-code", fRef, fNew]
 
 goldenGitDiff :: TestName -> FilePath -> IO ByteString -> TestTree
 goldenGitDiff name = goldenVsStringDiff name gitDiff
 
-testCommand :: String
-testCommand = "test-server"
+runSessionWithServer :: PluginDescriptor IdeState -> FilePath -> Session a -> IO a
+runSessionWithServer plugin = runSessionWithServer' [plugin] def def fullCaps
+
+runSessionWithServerFormatter :: PluginDescriptor IdeState -> String -> FilePath -> Session a -> IO a
+runSessionWithServerFormatter plugin formatter =
+  runSessionWithServer'
+    [plugin]
+    def {formattingProvider = T.pack formatter}
+    def
+    fullCaps
+
+-- | Run an action, with stderr silenced
+silenceStderr :: IO a -> IO a
+silenceStderr action = withTempFile $ \temp ->
+  bracket (openFile temp ReadWriteMode) hClose $ \h -> do
+    old <- hDuplicate stderr
+    buf <- hGetBuffering stderr
+    h `hDuplicateTo'` stderr
+    action `finally` do
+      old `hDuplicateTo'` stderr
+      hSetBuffering stderr buf
+      hClose old
+
+-- | Host a server, and run a test session on it
+-- Note: cwd will be shifted into @root@ in @Session a@
+runSessionWithServer' ::
+  -- | plugins to load on the server
+  [PluginDescriptor IdeState] ->
+  -- | lsp config for the server
+  Config ->
+  -- | config for the test session
+  SessionConfig ->
+  ClientCapabilities ->
+  FilePath ->
+  Session a ->
+  IO a
+runSessionWithServer' plugin conf sconf caps root s = silenceStderr $ do
+  (inR, inW) <- createPipe
+  (outR, outW) <- createPipe
+  -- restore cwd after running the session; otherwise the path to test data will be invalid
+  cwd <- getCurrentDirectory
+  server <-
+    async $
+      Ghcide.defaultMain
+        def
+          { argsHandleIn = pure inR,
+            argsHandleOut = pure outW,
+            argsDefaultHlsConfig = conf,
+            argsLogger = pure noLogging,
```

Comment:
Why `noLogging`?

Reply:
I think we need suppress the output from server otherwise the test report would be unreadable. And `silenceStderr` is used together with it, since there are many `hPutStrLn stderr` which don't comply with the logger.